File: euro_aip/euro_aip/briefing/models/route.py
"""Route data models."""

from dataclasses import dataclass, field
from datetime import datetime, timedelta
from typing import Optional, Tuple, List, Dict

from euro_aip.models.navpoint import NavPoint


@dataclass
class RoutePoint:
    """
    A point along a route with coordinates.

    Attributes:
        name: Waypoint name or ICAO code
        latitude: Latitude in decimal degrees
        longitude: Longitude in decimal degrees
        point_type: Type of point (departure, destination, alternate, waypoint)
    """
    name: str
    latitude: float
    longitude: float
    point_type: str = "waypoint"  # "departure", "destination", "alternate", "waypoint"

    @property
    def navpoint(self) -> NavPoint:
        """Get NavPoint for distance calculations."""
        return NavPoint(
            latitude=self.latitude,
            longitude=self.longitude,
            name=self.name
        )

    def to_dict(self) -> dict:
        """Serialize to dictionary."""
        return {
            'name': self.name,
            'latitude': self.latitude,
            'longitude': self.longitude,
            'point_type': self.point_type,
        }

    @classmethod
    def from_dict(cls, data: dict) -> 'RoutePoint':
        """Create from dictionary."""
        return cls(
            name=data['name'],
            latitude=data['latitude'],
            longitude=data['longitude'],
            point_type=data.get('point_type', 'waypoint'),
        )


@dataclass
class Route:
    """
    Flight route information with coordinates for spatial queries.

    Supports coordinate lookups for departure, destination, alternates,
    and waypoints. Integrates with NavPoint for distance calculations.

    Attributes:
        departure: Departure airport ICAO code
        destination: Destination airport ICAO code
        alternates: List of alternate airport ICAO codes
        waypoints: List of waypoint names along the route
    """
    departure: str
    destination: str
    alternates: List[str] = field(default_factory=list)
    waypoints: List[str] = field(default_factory=list)

    # Coordinates for spatial queries
    departure_coords: Optional[Tuple[float, float]] = None
    destination_coords: Optional[Tuple[float, float]] = None
    alternate_coords: Dict[str, Tuple[float, float]] = field(default_factory=dict)
    waypoint_coords: List[RoutePoint] = field(default_factory=list)

    # Flight details
    aircraft_type: Optional[str] = None
    departure_time: Optional[datetime] = None
    arrival_time: Optional[datetime] = None
    flight_level: Optional[int] = None
    cruise_altitude_ft: Optional[int] = None

    def get_all_airports(self) -> List[str]:
        """
        Get all airports involved in the route.

        Returns:
            List of unique ICAO codes for departure, destination, and alternates.
        """
        airports = [self.departure, self.destination]
        airports.extend(self.alternates)
        return list(dict.fromkeys(airports))  # Preserve order, remove duplicates

    def get_all_coordinates(self) -> List[Tuple[float, float]]:
        """
        Get all route coordinates for spatial queries.

        Returns ordered list of coordinates from departure through waypoints
        to destination.
        """
        coords = []
        if self.departure_coords:
            coords.append(self.departure_coords)
        coords.extend([(wp.latitude, wp.longitude) for wp in self.waypoint_coords])
        if self.destination_coords:
            coords.append(self.destination_coords)
        return coords

    def get_airport_coordinates(self) -> Dict[str, Tuple[float, float]]:
        """
        Get coordinates for all airports in route.

        Returns:
            Dict mapping ICAO codes to (lat, lon) tuples.
        """
        coords = {}
        if self.departure_coords:
            coords[self.departure] = self.departure_coords
        if self.destination_coords:
            coords[self.destination] = self.destination_coords
        coords.update(self.alternate_coords)
        return coords

    def get_route_navpoints(self) -> List[NavPoint]:
        """
        Get NavPoints for the full route.

        Returns:
            List of NavPoints from departure through waypoints to destination.
        """
        points = []
        if self.departure_coords:
            points.append(NavPoint(
                latitude=self.departure_coords[0],
                longitude=self.departure_coords[1],
                name=self.departure
            ))
        points.extend([wp.navpoint for wp in self.waypoint_coords])
        if self.destination_coords:
            points.append(NavPoint(
                latitude=self.destination_coords[0],
                longitude=self.destination_coords[1],
                name=self.destination
            ))
        return points

    def get_flight_window(self, buffer_minutes: int = 60) -> Tuple[datetime, datetime]:
        """
        Get the time window for the flight.

        Args:
            buffer_minutes: Buffer time after arrival (default 60 minutes)

        Returns:
            Tuple of (departure_time, arrival_time + buffer)

        Raises:
            ValueError: If departure_time is not set
        """
        if not self.departure_time:
            raise ValueError("Departure time not set")

        end_time = self.arrival_time or self.departure_time
        return (self.departure_time, end_time + timedelta(minutes=buffer_minutes))

    def to_dict(self) -> dict:
        """Serialize to dictionary."""
        return {
            'departure': self.departure,
            'destination': self.destination,
            'alternates': self.alternates,
            'waypoints': self.waypoints,
            'departure_coords': list(self.departure_coords) if self.departure_coords else None,
            'destination_coords': list(self.destination_coords) if self.destination_coords else None,
            'alternate_coords': {k: list(v) for k, v in self.alternate_coords.items()},
            'waypoint_coords': [wp.to_dict() for wp in self.waypoint_coords],
            'aircraft_type': self.aircraft_type,
            'departure_time': self.departure_time.isoformat() if self.departure_time else None,
            'arrival_time': self.arrival_time.isoformat() if self.arrival_time else None,
            'flight_level': self.flight_level,
            'cruise_altitude_ft': self.cruise_altitude_ft,
        }

    @classmethod
    def from_dict(cls, data: dict) -> 'Route':
        """Create from dictionary."""
        departure_time = None
        if data.get('departure_time'):
            departure_time = datetime.fromisoformat(data['departure_time'])

        arrival_time = None
        if data.get('arrival_time'):
            arrival_time = datetime.fromisoformat(data['arrival_time'])

        departure_coords = None
        if data.get('departure_coords'):
            departure_coords = tuple(data['departure_coords'])

        destination_coords = None
        if data.get('destination_coords'):
            destination_coords = tuple(data['destination_coords'])

        alternate_coords = {}
        if data.get('alternate_coords'):
            alternate_coords = {k: tuple(v) for k, v in data['alternate_coords'].items()}

        waypoint_coords = []
        if data.get('waypoint_coords'):
            waypoint_coords = [RoutePoint.from_dict(wp) for wp in data['waypoint_coords']]

        return cls(
            departure=data['departure'],
            destination=data['destination'],
            alternates=data.get('alternates', []),
            waypoints=data.get('waypoints', []),
            departure_coords=departure_coords,
            destination_coords=destination_coords,
            alternate_coords=alternate_coords,
            waypoint_coords=waypoint_coords,
            aircraft_type=data.get('aircraft_type'),
            departure_time=departure_time,
            arrival_time=arrival_time,
            flight_level=data.get('flight_level'),
            cruise_altitude_ft=data.get('cruise_altitude_ft'),
        )

    def __repr__(self) -> str:
        return f"Route({self.departure} -> {self.destination})"

File: Sources/RZFlight/Briefing/Route.swift
//
//  Route.swift
//  RZFlight
//
//  Flight route model for briefings.
//  Decodes from Python euro_aip.briefing JSON format.
//

import Foundation
import CoreLocation

/// A point along a flight route with coordinates.
public struct RoutePoint: Codable, Sendable {
    /// Waypoint name or ICAO code
    public let name: String

    /// Latitude in decimal degrees
    public let latitude: Double

    /// Longitude in decimal degrees
    public let longitude: Double

    /// Point type: "departure", "destination", "alternate", "waypoint"
    public let pointType: String

    /// Coordinates as CLLocationCoordinate2D
    public var coordinate: CLLocationCoordinate2D {
        return CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }

    enum CodingKeys: String, CodingKey {
        case name
        case latitude
        case longitude
        case pointType = "point_type"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.name = try container.decode(String.self, forKey: .name)
        self.latitude = try container.decode(Double.self, forKey: .latitude)
        self.longitude = try container.decode(Double.self, forKey: .longitude)
        self.pointType = try container.decodeIfPresent(String.self, forKey: .pointType) ?? "waypoint"
    }
}

/// Flight route information with coordinates for spatial queries.
///
/// Supports coordinate lookups for departure, destination, alternates,
/// and waypoints. Used for spatial NOTAM filtering.
public struct Route: Codable, Sendable {

    // MARK: - Core Fields

    /// Departure airport ICAO code
    public let departure: String

    /// Destination airport ICAO code
    public let destination: String

    /// Alternate airport ICAO codes
    public let alternates: [String]

    /// Waypoint names along the route
    public let waypoints: [String]

    // MARK: - Coordinates (stored as arrays for Codable)

    private let departureCoords: [Double]?
    private let destinationCoords: [Double]?
    private let alternateCoords: [String: [Double]]?

    /// Full waypoint data with coordinates
    public let waypointCoords: [RoutePoint]

    // MARK: - Flight Details

    /// Aircraft type code
    public let aircraftType: String?

    /// Planned departure time (UTC)
    public let departureTime: Date?

    /// Estimated arrival time (UTC)
    public let arrivalTime: Date?

    /// Cruise flight level (e.g., 350 for FL350)
    public let flightLevel: Int?

    /// Cruise altitude in feet
    public let cruiseAltitudeFt: Int?

    // MARK: - Computed Properties

    /// Departure airport coordinates
    public var departureCoordinate: CLLocationCoordinate2D? {
        guard let coords = departureCoords, coords.count == 2 else { return nil }
        return CLLocationCoordinate2D(latitude: coords[0], longitude: coords[1])
    }

    /// Destination airport coordinates
    public var destinationCoordinate: CLLocationCoordinate2D? {
        guard let coords = destinationCoords, coords.count == 2 else { return nil }
        return CLLocationCoordinate2D(latitude: coords[0], longitude: coords[1])
    }

    /// All airports involved in the route (unique)
    public var allAirports: [String] {
        var airports = [departure, destination]
        airports.append(contentsOf: alternates)
        // Remove duplicates while preserving order
        var seen = Set<String>()
        return airports.filter { seen.insert($0).inserted }
    }

    /// Get coordinates for all route points (departure through waypoints to destination)
    public var allCoordinates: [CLLocationCoordinate2D] {
        var coords: [CLLocationCoordinate2D] = []
        if let dep = departureCoordinate {
            coords.append(dep)
        }
        coords.append(contentsOf: waypointCoords.map { $0.coordinate })
        if let dest = destinationCoordinate {
            coords.append(dest)
        }
        return coords
    }

    /// Get coordinates for a specific airport
    public func coordinate(for icao: String) -> CLLocationCoordinate2D? {
        if icao == departure {
            return departureCoordinate
        }
        if icao == destination {
            return destinationCoordinate
        }
        if let altCoords = alternateCoords?[icao], altCoords.count == 2 {
            return CLLocationCoordinate2D(latitude: altCoords[0], longitude: altCoords[1])
        }
        return nil
    }

    /// Get the flight time window with buffer
    ///
    /// - Parameter bufferMinutes: Buffer time after arrival (default 60)
    /// - Returns: Tuple of (departure, arrival + buffer) or nil if departure time not set
    public func flightWindow(bufferMinutes: Int = 60) -> (start: Date, end: Date)? {
        guard let dep = departureTime else { return nil }
        let end = arrivalTime ?? dep
        return (dep, end.addingTimeInterval(Double(bufferMinutes) * 60))
    }

    // MARK: - CodingKeys

    enum CodingKeys: String, CodingKey {
        case departure
        case destination
        case alternates
        case waypoints
        case departureCoords = "departure_coords"
        case destinationCoords = "destination_coords"
        case alternateCoords = "alternate_coords"
        case waypointCoords = "waypoint_coords"
        case aircraftType = "aircraft_type"
        case departureTime = "departure_time"
        case arrivalTime = "arrival_time"
        case flightLevel = "flight_level"
        case cruiseAltitudeFt = "cruise_altitude_ft"
    }

    // MARK: - Decoding

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        self.departure = try container.decode(String.self, forKey: .departure)
        self.destination = try container.decode(String.self, forKey: .destination)
        self.alternates = try container.decodeIfPresent([String].self, forKey: .alternates) ?? []
        self.waypoints = try container.decodeIfPresent([String].self, forKey: .waypoints) ?? []

        self.departureCoords = try container.decodeIfPresent([Double].self, forKey: .departureCoords)
        self.destinationCoords = try container.decodeIfPresent([Double].self, forKey: .destinationCoords)
        self.alternateCoords = try container.decodeIfPresent([String: [Double]].self, forKey: .alternateCoords)
        self.waypointCoords = try container.decodeIfPresent([RoutePoint].self, forKey: .waypointCoords) ?? []

        self.aircraftType = try container.decodeIfPresent(String.self, forKey: .aircraftType)
        self.departureTime = try container.decodeIfPresent(Date.self, forKey: .departureTime)
        self.arrivalTime = try container.decodeIfPresent(Date.self, forKey: .arrivalTime)
        self.flightLevel = try container.decodeIfPresent(Int.self, forKey: .flightLevel)
        self.cruiseAltitudeFt = try container.decodeIfPresent(Int.self, forKey: .cruiseAltitudeFt)
    }
}

// MARK: - Memberwise Initializer

extension Route {
    /// Create a Route with all fields specified
    public init(
        departure: String,
        destination: String,
        alternates: [String] = [],
        waypoints: [String] = [],
        departureCoords: [Double]? = nil,
        destinationCoords: [Double]? = nil,
        alternateCoords: [String: [Double]]? = nil,
        waypointCoords: [RoutePoint] = [],
        aircraftType: String? = nil,
        departureTime: Date? = nil,
        arrivalTime: Date? = nil,
        flightLevel: Int? = nil,
        cruiseAltitudeFt: Int? = nil
    ) {
        self.departure = departure
        self.destination = destination
        self.alternates = alternates
        self.waypoints = waypoints
        self.departureCoords = departureCoords
        self.destinationCoords = destinationCoords
        self.alternateCoords = alternateCoords
        self.waypointCoords = waypointCoords
        self.aircraftType = aircraftType
        self.departureTime = departureTime
        self.arrivalTime = arrivalTime
        self.flightLevel = flightLevel
        self.cruiseAltitudeFt = cruiseAltitudeFt
    }
}

// MARK: - Protocol Conformances

extension Route: CustomStringConvertible {
    public var description: String {
        return "\(departure) -> \(destination)"
    }
}

extension RoutePoint: Hashable, Equatable, Identifiable {
    public var id: String { name }
}

extension RoutePoint {
    /// Create a RoutePoint with all fields specified
    public init(name: String, latitude: Double, longitude: Double, pointType: String = "waypoint") {
        self.name = name
        self.latitude = latitude
        self.longitude = longitude
        self.pointType = pointType
    }
}
